File: Routing/ActionRouter.swift
import Foundation
import os

/// The central coordinator between controller input and mode-specific actions.
///
/// `ActionRouter` sits between `ControllerManager` (raw button events) and
/// the `AppMode` conformers (button → action maps). Its responsibilities are:
///
/// 1. **Mode tracking** — updates the active mode when the frontmost app changes.
/// 2. **Combo detection** — checks for multi-button combos before single-button actions.
/// 3. **Action dispatch** — calls `action.perform()` for the resolved action.
///
/// All methods run on the main thread (enforced by `@MainActor`).
@MainActor
final class ActionRouter {

    // MARK: - Callbacks

    /// Called whenever the active mode changes. Use this to update the overlay
    /// and menu bar. Fires on the main thread.
    var onModeChanged: ((any AppMode) -> Void)?

    // MARK: - State

    /// The currently active mode. Changes when the frontmost app changes.
    private(set) var currentMode: any AppMode

    /// The set of buttons currently held down. Used for combo detection.
    private var heldButtons: Set<GamepadButton> = []

    // MARK: - Dependencies

    private let registry: ModeRegistry

    // MARK: - Initialisation

    init(registry: ModeRegistry) {
        self.registry   = registry
        self.currentMode = registry.fallback
    }

    // MARK: - Mode Switching

    /// Updates the active mode based on the given bundle identifier.
    ///
    /// If the new mode has the same `id` as the current mode, this is a no-op
    /// (no unnecessary overlay toasts or UI refreshes).
    func updateMode(for bundleIdentifier: String?) {
        let newMode = registry.mode(for: bundleIdentifier)
        guard newMode.id != currentMode.id else { return }

        Logger.actions.info("Mode changed: '\(self.currentMode.id)' → '\(newMode.id)'")
        currentMode = newMode
        onModeChanged?(newMode)
    }

    // MARK: - Button Dispatch

    /// Called by `ControllerManager` (via `AppDelegate`) when any button changes state.
    func buttonChanged(_ button: GamepadButton, isPressed: Bool) {
        if isPressed {
            heldButtons.insert(button)
            dispatch(pressed: button)
        } else {
            heldButtons.remove(button)
        }
    }

    // MARK: - Private: Dispatch Logic

    private func dispatch(pressed button: GamepadButton) {
        // 1. Check combos first. A combo fires only when the newly pressed
        //    button completes the set — all other buttons in the combo must
        //    already be held. Combos take priority over single-button actions.
        if let comboAction = resolveCombo(completedBy: button) {
            Logger.actions.debug("Firing combo action: '\(comboAction.id)'")
            comboAction.perform()
            return
        }

        // 2. Fall through to single-button action for this mode.
        if let action = currentMode.buttonMap[button] {
            Logger.actions.debug("Firing action: '\(action.id)' for button: '\(button.rawValue)'")
            action.perform()
        }
    }

    /// Returns the first combo in the current mode whose complete button set
    /// is satisfied by the current `heldButtons` state — or `nil` if none match.
    private func resolveCombo(completedBy triggerButton: GamepadButton) -> Action? {
        currentMode.comboMap.first { combo, _ in
            // The newly pressed button must be part of this combo...
            combo.buttons.contains(triggerButton) &&
            // ...and every button in the combo must now be held.
            combo.matches(heldButtons: heldButtons)
        }?.value
    }
}

// MARK: - ControllerManagerDelegate Conformance

extension ActionRouter: ControllerManagerDelegate {

    nonisolated func controllerDidConnect() {
        // Nothing to do — MenuBarController observes the connection state separately.
    }

    nonisolated func controllerDidDisconnect() {
        Task { @MainActor in
            heldButtons.removeAll()
        }
    }

    nonisolated func controllerButton(_ button: GamepadButton, didChange isPressed: Bool) {
        Task { @MainActor in
            buttonChanged(button, isPressed: isPressed)
        }
    }
}

// MARK: - AppDetectorDelegate Conformance

extension ActionRouter: AppDetectorDelegate {

    nonisolated func appDetector(_ detector: AppDetector, didDetectBundleIdentifier bundleIdentifier: String?) {
        Task { @MainActor in
            updateMode(for: bundleIdentifier)
        }
    }
}

File: pocket_study_remote/routing/action_router.py
"""
ActionRouter — the central coordinator between controller input and actions.

Responsibilities:
1. Tracks the active mode (updated by AppDetector via ``update_mode``).
2. Tracks which buttons are currently held (for combo detection).
3. Resolves button presses: checks combos first, falls through to
   single-button actions.
4. Fires the resolved action.

Threading:
    Button events arrive from the ControllerManager poll thread.
    App-change events arrive from the AppDetector poll thread.
    Both call into ActionRouter on their respective threads.

    ActionRouter uses a threading.Lock to protect ``_current_mode`` and
    ``_held_buttons``.  Action dispatch (``action.perform()``) is called
    while NOT holding the lock, so long-running actions (AppleScript) never
    block controller input.
"""

from __future__ import annotations

import logging
import threading
from typing import Callable

from ..core.app_mode import AppMode
from ..core.button_combo import ButtonCombo
from ..core.gamepad_button import GamepadButton
from ..core.mode_registry import ModeRegistry

logger = logging.getLogger(__name__)


class ActionRouter:
    """
    Routes button presses to actions based on the active mode.

    Args:
        registry:        Populated ``ModeRegistry``.
        on_mode_changed: Optional callback fired (with the new mode) whenever
                         the active mode changes.  Use this to update the
                         status bar label and trigger the overlay.
    """

    def __init__(
        self,
        registry: ModeRegistry,
        on_mode_changed: Callable[[AppMode], None] | None = None,
    ) -> None:
        self._registry        = registry
        self._on_mode_changed = on_mode_changed
        self._lock            = threading.Lock()
        self._current_mode: AppMode           = registry.fallback
        self._held_buttons:  set[GamepadButton] = set()

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    @property
    def current_mode(self) -> AppMode:
        with self._lock:
            return self._current_mode

    # ------------------------------------------------------------------
    # Called by AppDetector
    # ------------------------------------------------------------------

    def update_mode(self, bundle_identifier: str | None) -> None:
        """Switch to the mode registered for ``bundle_identifier``."""
        new_mode = self._registry.mode_for(bundle_identifier)

        with self._lock:
            if new_mode.id == self._current_mode.id:
                return
            self._current_mode = new_mode
            self._held_buttons.clear()   # stale held state is meaningless after a mode switch

        logger.info("Mode changed → %s", new_mode.display_name)

        if self._on_mode_changed:
            self._on_mode_changed(new_mode)

    # ------------------------------------------------------------------
    # Called by ControllerManager
    # ------------------------------------------------------------------

    def button_changed(self, button: GamepadButton, is_pressed: bool) -> None:
        """Process a button press or release."""
        action = None

        with self._lock:
            if is_pressed:
                self._held_buttons.add(button)
                action = self._resolve(button)
            else:
                self._held_buttons.discard(button)

        # Fire outside the lock so AppleScript/keystroke calls don't block input.
        if action is not None:
            action()

    # ------------------------------------------------------------------
    # Private: dispatch logic
    # ------------------------------------------------------------------

    def _resolve(self, pressed: GamepadButton):
        """
        Return the action to fire for the newly pressed button.

        Combos are checked first.  A combo fires only when the pressed
        button completes its full button set — all other members must
        already be held.  If no combo matches, return the single-button
        action (or None if unassigned).
        """
        mode = self._current_mode
        held = self._held_buttons   # already includes ``pressed``

        # 1. Check combos
        for combo, action in mode.combo_map.items():
            if pressed in combo.buttons and combo.matches(held):
                logger.debug("Combo fired: %s → %s", combo.buttons, action.id)
                return action

        # 2. Single-button action
        action = mode.button_map.get(pressed)
        if action:
            logger.debug("Button fired: %s → %s", pressed.display_name, action.id)
        return action
